File: src/main/java/com/refinedmods/refinedstorage/network/grid/GridTransferMessage.java
package com.refinedmods.refinedstorage.network.grid;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.api.network.grid.GridType;
import com.refinedmods.refinedstorage.api.network.grid.IGrid;
import com.refinedmods.refinedstorage.container.GridContainerMenu;
import com.refinedmods.refinedstorage.util.StackUtils;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import net.neoforged.neoforge.network.handling.PlayPayloadContext;

import java.util.List;

public class GridTransferMessage implements CustomPacketPayload {
    public static final ResourceLocation ID = new ResourceLocation(RS.ID, "grid_transfer");

    private final ItemStack[][] recipe;

    public GridTransferMessage(final ItemStack[][] recipe) {
        this.recipe = recipe;
    }

    public static GridTransferMessage decode(FriendlyByteBuf buf) {
        int slots = buf.readInt();
        final ItemStack[][] recipe = new ItemStack[slots][];
        for (int i = 0; i < slots; i++) {
            int numberOfIngredients = buf.readInt();
            recipe[i] = new ItemStack[numberOfIngredients];
            for (int j = 0; j < numberOfIngredients; j++) {
                recipe[i][j] = StackUtils.readItemStack(buf);
            }
        }
        return new GridTransferMessage(recipe);
    }

    public static void handle(GridTransferMessage message, PlayPayloadContext ctx) {
        ctx.player().ifPresent(player -> ctx.workHandler().submitAsync(() -> {
            if (player.containerMenu instanceof GridContainerMenu) {
                IGrid grid = ((GridContainerMenu) player.containerMenu).getGrid();

                if (grid.getGridType() == GridType.CRAFTING || grid.getGridType() == GridType.PATTERN) {
                    grid.onRecipeTransfer(player, message.recipe);
                }
            }
        }));
    }

    @Override
    public void write(FriendlyByteBuf buf) {
        buf.writeInt(recipe.length);
        for (ItemStack[] stacks : recipe) {
            buf.writeInt(stacks.length);
            for (ItemStack possibleStack : stacks) {
                StackUtils.writeItemStack(buf, possibleStack);
            }
        }
    }

    @Override
    public ResourceLocation id() {
        return ID;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/grid/GridTransferMessage.kt
package com.refinedmods.refinedstorage.network.grid

import com.refinedmods.refinedstorage.api.network.grid.GridType
import com.refinedmods.refinedstorage.container.GridContainer
import com.refinedmods.refinedstorage.util.StackUtils.readItemStack
import com.refinedmods.refinedstorage.util.StackUtils.writeItemStack
import mezz.jei.api.gui.ingredient.IGuiIngredient
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.container.Slot
import net.minecraft.item.ItemStack
import net.minecraft.network.PacketByteBuf
import net.minecraftforge.fml.network.NetworkEvent
import java.util.*
import java.util.function.Supplier

class GridTransferMessage {
    private var inputs: Map<Int, IGuiIngredient<ItemStack>?>? = null
    private var slots: List<Slot>? = null
    private val recipe = arrayOfNulls<Array<ItemStack?>?>(9)

    constructor() {}
    constructor(inputs: Map<Int?, IGuiIngredient<ItemStack?>?>, slots: List<Slot>?) {
        this.inputs = inputs
        this.slots = slots
    }

    companion object {
        fun decode(buf: PacketByteBuf): GridTransferMessage {
            val msg = GridTransferMessage()
            val slots: Int = buf.readInt()
            for (i in 0 until slots) {
                val ingredients: Int = buf.readInt()
                msg.recipe[i] = arrayOfNulls(ingredients)
                for (j in 0 until ingredients) {
                    msg.recipe[i]!![j] = readItemStack(buf)
                }
            }
            return msg
        }

        fun encode(message: GridTransferMessage, buf: PacketByteBuf) {
            buf.writeInt(message.slots!!.size)
            for (slot in message.slots!!) {
                val ingredient: IGuiIngredient<ItemStack>? = message.inputs!![slot.getSlotIndex() + 1]
                val ingredients: MutableList<ItemStack> = ArrayList()
                if (ingredient != null) {
                    for (possibleStack in ingredient.getAllIngredients()) {
                        if (possibleStack != null) {
                            ingredients.add(possibleStack)
                        }
                    }
                }
                buf.writeInt(ingredients.size)
                for (possibleStack in ingredients) {
                    writeItemStack(buf, possibleStack)
                }
            }
        }

        fun handle(message: GridTransferMessage, ctx: Supplier<NetworkEvent.Context>) {
            val player: PlayerEntity = ctx.get().getSender()
            if (player != null) {
                ctx.get().enqueueWork({
                    if (player.openContainer is GridContainer) {
                        val grid = (player.openContainer as GridContainer).grid
                        if (grid!!.gridType === GridType.CRAFTING || grid!!.gridType === GridType.PATTERN) {
                            grid!!.onRecipeTransfer(player, message.recipe)
                        }
                    }
                })
            }
            ctx.get().setPacketHandled(true)
        }
    }
}